proposed/2023/Add-dotnet-search.md

```diff
@@ -0,0 +1,107 @@
+<!-- Replace `Title` with an appropriate title for your design -->
+
+- [Nigusu](https://github.com/Nigusu-Allehu) <!-- GitHub username link -->
+- [GitHub Issue](https://github.com/NuGet/Home/issues/6060) <!-- GitHub Issue link -->
+
+## Summary
+
+<!-- One-paragraph description of the proposal. -->
+The specification seeks to add `nuget search` functionality to dotnet. The functionality will be added to dotnet as a search command: `dotnet package search`.  
+
+## Motivation
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+As outlined in the customer request on this GitHub [issue](https://github.com/NuGet/Home/issues/6060), there is a call for the incorporation of `nuget list` functionality into dotnet. The customers are requesting for this feature, but the `search` command can replicate the same functionality, with the only difference being the quantity of package results it outputs. The search command mitigates unnecessary server use by limiting the number of results through the `--take` option. By integrating this search functionality, we aim to accommodate our customers' needs, especially for scripting purposes.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+Imagine you have your dotnet CLI open. You would like to look up a NuGet package named `MyPackage` from a source `<MySource>`. No worries, you can just use the following command : `dotnet package search MyPackage --source <MySource>`. It will provide you with the list of the packages in the source `<MySource>` that match with the search criteria.
+
+### Technical explanation
+
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+The `package search [search terms] [options]` command will have the following options
+
+| Option | Function |
+|---------|:----------|
+| `--source` | A list of packages to search |
+| `--exact-match` | Return exact matches only as a search result |
+| `--prerelease` | Allow prerelease packages to be shown. |
+| `--non-interactive` | Do not prompt for user input or confirmations.|
+| `--take` | The number of results to return. The default value is 20.|
+| `--help` | Show command help and usage information |
+| `--verbosity` | Display the amount of details in the output: normal, quiet, detailed. |
+|||
+
+#### **Algorithm**
+
+1. Read and parse arguments from user
+2. Use the nuget.exe Search API to look for the specific term in the provided source. The search filter will be composed of the `--source`, `[Search term]`, `--prerelease` values.
+3. Based on `--verbosity` value, provide with a list of packages with various verbosity
```

Comment:
Some tools, like [`git log`, allow you to pass a format string](https://www.git-scm.com/docs/git-log#_commit_formatting), so you can format it in any way you want. In theory, you can probably get it to generate (almost) JSON.

I don't want to over complicate things, and I know this will feel hard because it's a pattern we don't already have anywhere else in NuGet, but should we consider removing `--verbosity` and do some kind of `--format "\i | \v | \o\n\d"` ? 

We'd need to come up with a format code for every metadata item, but then customers could possibly do something like ``$json = '[' + `dotnet package search sql --format "{\"id\": \"\\i\", \"version\": \"\\v\", \"description\": \"\\d\" }"` + ']'`` and then process it as json. We'd also need to think a bit harder about what other escaping we'd need, this was just a 5 second example. For example, we'd probably need two copies of all properties that have free text, just in case the metadata contains a double quotes character that would otherwise break json. For example, `\d` could output the package description as-is, and `\D` outputs the package description with json encoding.

Since we're a .NET project, we could consider more `string.Format` style syntax, like ``--format "{id} | {version} | {downloads}\n{description}"`` and ``--format "\{ \"id\": \"{id}\", \"version\": \"{version}\", \"description\": \"{description:j}\" \}"``, which is clearer to people not familar with C's `printf` style formatting, but much, much more verbose.

My concern is that if we start off with `--verbosity`, but then later get requests to add something like `--format`, then there might be complicated questions or implementation details around what happens when `--format` and `--verbosity` are provided at the same time. Although a simple answer would be to error out when both are provided at the same time, commands that can't be used in a wrong way are generally better than commands that can. Hence, why I think it's worthwhile to think about whether this is something we would want to do in a V1.  It would also be a great feature to encourage customers to get off `nuget.exe search` and onto `dotnet package search`

I'm really not sure how complex this might be to implement. Might be worth doing a 1 day spike to see how close we can get a string format parser working, to figure out if this suggestion is feasible?

Reply:
Maybe we propose no `--verbosity` in v1?

TBH, I don't haste just having 1 verbosity, especially since adding more things will be easier than taking away things

Reply:
I like this idea. I think `--format` would be more useful and practical to users. I will have just one verbosity, `--Normal`, when the `--format` is not specified. I am thinking of, for this PR, removing `--verbosity`. And then adding `--format` in its own spec PR and we can discuss its syntax in that PR.